### PR TITLE
Update all_to_all op interpreter to fix issue#2433

### PR DESCRIPTION
**Pull request by abhigunj**
fixes: https://github.com/openxla/stablehlo/issues/2433

```
scattered_parts...@receiver = [split_parts...@sender[receiver_index] for sender in process_group] where receiver_index = process_group.index(receiver).
```
each `sender/process` from `process_group` contributes a `split_part` to the `scattered_parts`. This PR ensures the order of `split_part` in `scattered_parts` matches with the order of the `sender/processes` within the `process_group`.

### CODE REVIEW

File: stablehlo/reference/Ops.cpp [hunk 1]
```diff
@@ -1198,20 +1198,27 @@ Tensor allToAllOp(const Tensor &operand, Axis splitDimension,
...
 
   auto processGroup = processGroups.findGroup(process->getId());
-  if (!processGroup)
+  if (!processGroup.has_value())
     llvm::report_fatal_error(invalidArgument(
         "Failed to find process group with process_id: (%d, %d)",
         process->getId().replicaId, process->getId().partitionId));
-  auto groupOperands = process->rendezvous(*processGroup, channelId, {operand})
-                           .getSortedTensors();
+  auto rendezvousResult =
```
Comment (ghpvnist) [inline stablehlo/reference/Ops.cpp:1205]: Implementation looks good. Suggestion to simplify code and make the variable names better match the spec:
```
auto operands = process->rendezvous(*processGroup, channelId, {operand});
auto receiver_index =
    llvm::find(*processGroup, process->getId()) - processGroup->begin();

SmallVector<Tensor> scatteredParts;
for (const auto& sender : *processGroup) {
  auto splitParts = split(operands.lookup(sender).front(), splitCount,
                          splitDimension, operand.getType().getContext());
  scatteredParts.push_back(splitParts[receiver_index]);
}
```

File: stablehlo/tests/interpret/all_to_all.mlir [hunk 2]
```diff
@@ -32,6 +32,38 @@ module @cross_replica {
 
 // -----
 
+module @cross_replica_reverse_replica_groups {
```
Comment (ghpvnist) [inline stablehlo/tests/interpret/all_to_all.mlir:35]: Let's rename it to: `@cross_replica_issue_2433`
precedent in https://github.com/openxla/stablehlo/blob/eba821aa1c54a21d70331d7926dfc8b929f988f3/stablehlo/tests/interpret/all_gather.mlir#L52.

**Review by ghpvnist — APPROVED** [review]
LGTM with minor touch ups. Thanks @Hatsunespica for the bug and @abhigunj for the fix!

File: stablehlo/reference/Ops.cpp (hunk 1 above)
Comment (abhigunj) [reply stablehlo/reference/Ops.cpp]: agree. Thanks for the review. updated the PR with some modifications. 

renamed `operands` -> `groupOperands` to make it future proof for variadic operands.
 `*processGroup`  ->`processGroup.value()`

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -1198,20 +1198,21 @@ Tensor allToAllOp(const Tensor &operand, Axis splitDimension,
...
   auto processGroup = processGroups.findGroup(process->getId());
-  if (!processGroup)
+  if (!processGroup.has_value())
     llvm::report_fatal_error(invalidArgument(
         "Failed to find process group with process_id: (%d, %d)",
         process->getId().replicaId, process->getId().partitionId));
-  auto groupOperands = process->rendezvous(*processGroup, channelId, {operand})
-                           .getSortedTensors();
+  auto groupOperands =
+      process->rendezvous(processGroup.value(), channelId, {operand});
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:1206]: nit:Can use `*processGroup`? here and below?

**Review by sdasgup3 — APPROVED** [review]
lgtm with a nit.

**Comment by GleasonK** [comment]
LGTM, no need for approval from me on this one, other reviewers are enough

File: stablehlo/reference/Ops.cpp (hunk 3 above)
Comment (abhigunj) [reply stablehlo/reference/Ops.cpp:1206]: any advantage of using *? `.value()` also returns reference to the object at std::optional and throws exception for null values, I think it is preferred  vs using * here?  also, it has better readability imo.

Comment (abhigunj) [reply stablehlo/reference/Ops.cpp:1206]: sg, thanks for the review.

**Pull request merged**